shivyc/il_cmds/math.py

```diff
@@ -118,6 +118,70 @@ class Mult(_AddMult):
     Inst = asm_cmds.Imul
 
 
+class _BitShiftCmd(ILCommand):
+    """Base class for bitwise shift commands."""
+
+    # The ASM instruction to generate for this command. Override this value
+    # in subclasses.
+    Inst = None
+
+    def __init__(self, output, arg1, arg2): # noqa D102
+        self.output = output
+        self.arg1 = arg1
+        self.arg2 = arg2
+
+    def inputs(self): # noqa D102
+        return [self.arg1, self.arg2]
+
+    def outputs(self): # noqa D102
+        return [self.output]
+
+    def abs_spot_pref(self): # noqa D102
+        return {self.arg2: [spots.RCX]}
+
+    def rel_spot_pref(self): # noqa D102
+        return {self.output: [self.arg1]}
+
+    def make_asm(self, spotmap, home_spots, get_reg, asm_code): # noqa D102
+        arg1_spot = spotmap[self.arg1]
+        arg1_size = self.arg1.ctype.size
+        arg2_spot = spotmap[self.arg2]
+        arg2_size = self.arg2.ctype.size
+
+        # According Intel® 64 and IA-32 software developer's manual
+        # Vol. 2B 4-582 second (count) operand must be represented as
+        # imm8 or CL register.
+        if not self._is_imm8(arg2_spot) and arg2_spot != spots.RCX:
+            asm_code.add(asm_cmds.Mov(spots.RCX, arg2_spot, arg2_size))
+            arg2_spot = spots.RCX
+
+        if spotmap[self.output] == arg2_spot:
+            asm_code.add(self.Inst(arg1_spot, arg2_spot, arg1_size, 1))
+        else:
+            out_spot = spotmap[self.output]
+            temp_spot = get_reg([out_spot, arg1_spot], [arg2_spot])
+            asm_code.add(asm_cmds.Mov(temp_spot, arg1_spot, arg1_size))
```

Comment:
You can add a check here that outputs this `Mov` only if it is necessary (i.e. only if `arg1_spot != temp_spot`), just like you have below